### PR TITLE
Passing all Travis tests

### PR DESCRIPTION
**Pull request by CBreshears**
Whipped this up into shape to pass the Travis tests posted by the original repository. 

This version adds a "memory_stream" data type to allow images to be compressed/decompressed from/to memory locations. This allows ArrayDS to use and handle the file writing/reading when dealing with images (and ImageDS).

### CODE REVIEW

File: src/lib/openjp2/image.c [hunk 1]
```diff
@@ -48,8 +48,8 @@ opj_image_t* OPJ_CALLCONV opj_image_create(OPJ_UINT32 numcmpts,
         image->color_space = clrspc;
         image->numcomps = numcmpts;
         /* allocate memory for the per-component information */
-        image->comps = (opj_image_comp_t*)opj_calloc(image->numcomps,
-                       sizeof(opj_image_comp_t));
+        image->comps = (opj_image_comp_t*)opj_calloc(1,
+                       image->numcomps * sizeof(opj_image_comp_t));
```
Comment (nalinigans) [inline src/lib/openjp2/image.c:52]: Is this change necessary? I did keep the original code as much as possible, especially if the effect is the same.

File: src/lib/openjp2/cio.c [hunk 2]
```diff
@@ -204,6 +211,7 @@ void OPJ_CALLCONV opj_stream_destroy(opj_stream_t* p_stream)
         if (l_stream->m_free_user_data_fn) {
             l_stream->m_free_user_data_fn(l_stream->m_user_data);
         }
+//else {fprintf(stderr, "ERROR: No m_free_user_data_fn function set\n");}
```
Comment (nalinigans) [inline src/lib/openjp2/cio.c:214]: Would it be possible to make this a true error? Or is the being kept for future debugging? For future debugging, you could use #ifdef DEBUG...

File: src/lib/openjp2/cio.c [hunk 3]
```diff
@@ -154,6 +154,13 @@ void opj_read_float_LE(const OPJ_BYTE * p_buffer, OPJ_FLOAT32 * p_value)
         *(l_data_ptr--) = *(p_buffer++);
     }
 }
+/** CPB *******************************************************************/
+/* Added for Omics Data Automation                                        */
+/**************************************************************************/
+
+/**************************************************************************/
+/* End Additions for Omics Data Automation                                */
+/** CPB END ***************************************************************/
```
Comment (nalinigans) [inline src/lib/openjp2/cio.c:163]: There seem to be no additions here. Is that intended?

File: src/lib/openjp2/cio.c [hunk 4]
```diff
@@ -264,6 +272,11 @@ void OPJ_CALLCONV opj_stream_set_user_data(opj_stream_t* p_stream,
     if (!l_stream) {
         return;
     }
+    /* CPB: if called after default chunk of memory allocated */
+    if (l_stream->m_user_data) {
+        p_function(l_stream->m_user_data);
```
Comment (nalinigans) [inline src/lib/openjp2/cio.c:277]: What is the purpose of p_function()?

File: src/lib/openjp2/j2k.c [hunk 5]
```diff
@@ -12017,34 +12003,25 @@ static OPJ_BOOL opj_j2k_init_info(opj_j2k_t *p_j2k,
...
     /*l_cstr_info->numcomps = p_j2k->m_image->numcomps;
-
     l_cstr_info->numlayers = (&p_j2k->m_cp.tcps[0])->numlayers;
-
     l_cstr_info->numdecompos = (OPJ_INT32*) opj_malloc(p_j2k->m_image->numcomps * sizeof(OPJ_INT32));
-
     for (compno=0; compno < p_j2k->m_image->numcomps; compno++) {
             l_cstr_info->numdecompos[compno] = (&p_j2k->m_cp.tcps[0])->tccps->numresolutions - 1;
     }
-
```
Comment (nalinigans) [inline src/lib/openjp2/j2k.c:12047]: Seems to be no change in this file other than removing lines. It's best to keep these type of changes minimal as it will may make future merging with the original master difficult.

File: src/lib/openjp2/jp2.c [hunk 6]
```diff
@@ -1079,7 +1079,7 @@ static OPJ_BOOL opj_jp2_apply_pclr(opj_image_t *image,
 
         /* Palette mapping: */
         new_comps[i].data = (OPJ_INT32*)
-                            opj_image_data_alloc(sizeof(OPJ_INT32) * old_comps[cmp].w * old_comps[cmp].h);
+                            opj_image_data_alloc(old_comps[cmp].w * old_comps[cmp].h * sizeof(OPJ_INT32));
```
Comment (nalinigans) [inline src/lib/openjp2/jp2.c:1082]: Do we need this re-arranging of multiplication operations here?

File: src/lib/openmj2/j2k.c [hunk 7]
```diff
@@ -331,7 +331,7 @@ static int j2k_calculate_tp(opj_cp_t *cp, int img_numcomp, opj_image_t *image,
 
     OPJ_ARG_NOT_USED(img_numcomp);
 
-    j2k->cur_totnum_tp = (int *) opj_malloc(sizeof(int) * cp->tw * cp->th);
+    j2k->cur_totnum_tp = (int *) opj_malloc(cp->tw * cp->th * sizeof(int));
```
Comment (nalinigans) [inline src/lib/openmj2/j2k.c:334]: I think I am missing something with the rearrangement of operations .

**Review by nalinigans — APPROVED** [review]
Looks good to me. Just minor comments if you want to address them that is. I am a little confused by the two PRs. Does this supersede the previous one PR #1?

File: src/lib/openmj2/j2k.c (hunk 7 above)
Comment (CBreshears) [reply src/lib/openmj2/j2k.c:334]: Again, just me trying to use consistent style. No need to rearrange.

File: src/lib/openjp2/jp2.c (hunk 6 above)
Comment (CBreshears) [reply src/lib/openjp2/jp2.c:1082]: Just me trying to use consistent style. No need to rearrange.

File: src/lib/openjp2/j2k.c (hunk 5 above)
Comment (CBreshears) [reply src/lib/openjp2/j2k.c:12047]: OK. Just my OCD about blank line in code.

File: src/lib/openjp2/cio.c (hunk 4 above)
Comment (CBreshears) [reply src/lib/openjp2/cio.c:277]: This frees up memory. The l_stream->m_user_data is assumed to be NULL, if not, free up whatever was left over using the appropriate function (as passed in through parameter).

File: src/lib/openjp2/cio.c (hunk 3 above)
Comment (CBreshears) [reply src/lib/openjp2/cio.c:163]: Originally that did have stuff. Moved all that off to other files or places in source file. Not needed now.

File: src/lib/openjp2/cio.c (hunk 2 above)
Comment (CBreshears) [reply src/lib/openjp2/cio.c:214]: Yes, that looks like an old debugging thing. Probably isn't needed since m_free_user_data_fn() is set by default on creation of l_stream object.

File: src/lib/openjp2/image.c (hunk 1 above)
Comment (CBreshears) [reply src/lib/openjp2/image.c:52]: Not necessary; not sure why I changed this. What's the quick fix to revert back?

**Comment by nalinigans** [comment]
Do we still need PR #1?

**Comment by CBreshears** [comment]
No, the first PR was a premature attempt. Cancel or clobber or kill or whatever should be done to remove it.
